### PR TITLE
fix: empty data OnProgress by race conditions of AsyncProgressWorker

### PR DESCRIPTION
**Pull request by legendecas**
Fixes https://github.com/nodejs/node-addon-api/issues/821

The callback of ThreadSafeFunction is not been invoked immediately on the callback of uv_async_t (uv io poll), rather the callback of TSFN is invoked on the right next uv idle callback. There are chances that during the deferring the signal of uv_async_t is been sent again, i.e. potential not coalesced two calls of the TSFN callback.

However, when the race condition happens, the AsyncProgressWorker::OnProgress is likely not been called yet. There is hardly a fix on ensuring the race condition not happening. 

In summary, the fix is suppressing the unexpected duplicated calls of AsyncProgressWorker::OnProgress.

### CODE REVIEW

**Review by mhdawson — APPROVED** [review]
LGTM

**Review by gabrielschulhof — APPROVED** [review]

**Comment by mhdawson** [comment]
@legendecas I tried to land this and had 2 issues

1) it reported an unused value warning on line 90 of test/asyncprogressworker.cc.  I update the line so not an issue:
-  void OnProgress(const ProgressData* data, size_t count) override {
+  void OnProgress(const ProgressData*, size_t count) override {

2) The tests seem to stop after the asyncprogressworker test:

```shell
napiVersion:6
napiVersion:6
Testing with N-API Version '6'.
Starting test suite

Running test 'addon_build'
   >Preparing native addons to build
   >Building addon: 'echo addon'
   >Running test for: 'echo addon'
   >Building addon: 'echo-addon'
   >Running test for: 'echo-addon'
Running test 'addon'
Running test 'addon_data'
Running test 'arraybuffer'
Running test 'asynccontext'
Running test 'asyncprogressqueueworker'
Running test 'asyncprogressworker'
```

We'll need to figure that out before this can land.

**Review by NickNaso — APPROVED** [review]
LGTM

**Comment by legendecas** [comment]
> @mhdawson  The tests seem to stop after the asyncprogressworker test

Yeah, I've identified the problem. There might be an exception leaking in threadsafe function callbacks in node core so that the promise chains of the node-addon-api are broke up yet the thrown JavaScript exception is not handled properly and the process exits silently. I'll try to file a minimum reproduce to get that fixed in node core. Besides that, I'll update the PR to get tests running.

**Comment by legendecas** [comment]
@mhdawson GitHub Actions are up and all tests get passed :) PTAL

**Comment by legendecas** [comment]
> > @mhdawson  The tests seem to stop after the asyncprogressworker test
> 
> Yeah, I've identified the problem. There might be an exception leaking in threadsafe function callbacks in node core so that the promise chains of the node-addon-api are broke up yet the thrown JavaScript exception is not handled properly and the process exits silently. I'll try to file a minimum reproduce to get that fixed in node core. Besides that, I'll update the PR to get tests running.

Just filed WIP https://github.com/nodejs/node/pull/36510. Need to figure out if there are any other cases that also swallowing exceptions.

**Comment by mhdawson** [comment]
@legendecas thanks for figuring out the issue and fixing up the tests. They all pass  for me, landing.

**Comment by mhdawson** [comment]
Landed as 86feeebf5489126ab941afc92866eba0257ff8d2
